yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts

```diff
@@ -0,0 +1,121 @@
+// This is set to a larger value than MAX_PRIVATE_LOGS_PER_TX (currently 32) because there could be more than
+// MAX_PRIVATE_LOGS_PER_TX indexes consumed in case the logs are squashed. This happens when the log contains a note
+// and the note is nullified in the same tx.
+// Note: Set it to 95 because `e2e_pending_note_hashes_contract` test hit 95 indexes emitted and 100 makes
+// `docs_examples.test.ts` fail as our JSON RPC doesn't seem to allow arrays with length larger than 99. I (benesjan)
+// think this window length should be fine.
+export const WINDOW_LEN = 95;
+
+/**
+ * Syncs the highest finalized tagging index and pending tagging indexes for a given secret.
+ * @param secret - The secret that's unique for (sender, recipient, contract) tuple while the direction of
+ * sender -> recipient matters.
+ * @param app - The address of the contract that the logs are tagged for. Needs to be provided because we perform
+ * second round of siloing in this function which is necessary because kernels do it as well (they silo first field
+ * of the private log which corresponds to the tag).
+ * @remarks When syncing the indexes as sender we don't care about the log contents - we only care about the highest
+ * pending and highest finalized indexes as that guides the next index choice when sending a log. The next index choice
+ * is simply the highest pending index plus one (or finalized if pending is undefined).
+ */
+export async function syncSenderTaggingIndexes(
+  secret: DirectionalAppTaggingSecret,
+  app: AztecAddress,
+  aztecNode: AztecNode,
+  taggingDataProvider: SenderTaggingDataProvider,
+): Promise<void> {
+  // # Explanation of how syncing works
+  //
+  // When choosing an index, we select: highest pending index + 1 (or highest finalized index + 1 if no pending).
+  // If the chosen index is more than WINDOW_LEN from the highest finalized index, we throw an error. By having this
+  // hard limit we give a guarantee to a recipient that he doesn't need to look further than WINDOW_LEN ahead of the
+  // highest finalized index.
+  //
+  // This function synchronizes these two values by iteratively querying the node for a window of indexes at a time,
+  // storing all those indexes as pending, and then checking the status of each pending index to update its
+  // finalization status accordingly.
+
+  const finalizedIndex = await taggingDataProvider.getLastFinalizedIndex(secret);
+
+  let start = finalizedIndex === undefined ? 0 : finalizedIndex + 1;
+  let end = start + WINDOW_LEN;
+
+  let previousFinalizedIndex = finalizedIndex;
+  let newFinalizedIndex = undefined;
+
+  while (true) {
+    // Load and store indexes for the current window. These indexes may already exist in the database if txs using
+    // them were previously sent from this PXE. Any duplicates are handled by the tagging data provider.
+    await loadAndStoreNewTaggingIndexes(secret, app, start, end, aztecNode, taggingDataProvider);
```

Comment:
A broad-strokes explanation of wtf is happening would be extremely useful. My tldr is something like

- we search for all used tags from the last tip into the future
- we check if any of the tags we found was finalized, which moves the tip forward
- repeat until the tip doesnt move forward
- record the information we found

This nicely showcases that moving the window forward requires 2 queries (logs + tx receipt), so if going from e.g. index 0 to 500 with a window length of 100, that'll take 10 roundrips (5*2) at least.